jwave/geometry.py

```diff
@@ -461,10 +463,102 @@ def __call__(self, p: Field, u: Field, rho: Field):
             return p.on_grid[self.positions[0]]
         elif len(self.positions) == 2:
             return p.on_grid[self.positions[0],
-                             self.positions[1]]    # type: ignore
+            self.positions[1]]  # type: ignore
         elif len(self.positions) == 3:
             return p.on_grid[self.positions[0], self.positions[1],
-                             self.positions[2]]    # type: ignore
+            self.positions[2]]  # type: ignore
+        else:
+            raise ValueError(
+                "Sensors positions must be 1, 2 or 3 dimensional. Not {}".
+                format(len(self.positions)))
+
+
+def _bli_function(x0, x, n, include_imag=False):
+    # x0 = position of sensors along axis
+    # x = grid
+    # n = size of grid
+    # include_imag = include the imaginary component?
+    dx = jnp.where((x - x0[:, None]) == 0, x - x0[:, None] + 1e-4, x - x0[:, None])  # Stop us getting nans.
+    dx_nonzero = (x - x0[:, None]) != 0
+
+    if n % 2 == 0:
+        y = jnp.sin(jnp.pi * dx) / \
+            jnp.tan(jnp.pi * dx / n) / n
+        y -= jnp.sin(jnp.pi * x0[:, None]) * jnp.sin(jnp.pi * x) / n
+        if include_imag:
+            y += 1j * jnp.cos(jnp.pi * x0[:, None]) * jnp.sin(jnp.pi * x) / n
+    else:
+        y = jnp.sin(jnp.pi * dx) / \
+            jnp.sin(jnp.pi * dx / n) / n
+
+    # Deal with case of precisely on grid.
+    y = y * jnp.all(dx_nonzero, axis=1)[:, None] + (1 - dx_nonzero) * (~jnp.all(dx_nonzero, axis=1)[:, None])
+    return y
+
+
+@register_pytree_node_class
+class BLISensors:
+    """ Band-limited interpolant (off-grid) sensors.
+
+    Args:
+        positions (Tuple of List of float): The positions of the sensors.
+        n (Tuple of int): The shape of the grid.
+    """
+
+    positions: Tuple[ArrayLike]
+
```

Comment:
Since it seems that you are setting `self.N` in the `__init__` function, it may be worth defining the type of `N` here, in the same way as it is done for `positions`